docs/coding-guidelines/vectorization-guidelines.md

```diff
@@ -0,0 +1,1056 @@
+
+In recent releases, .NET has introduced many new APIs for vectorization. The vast majority of them are hardware specific, so they require users to provide an implementation per processor architecture (x64 and/or arm64), with the option of using the most optimal instructions for hardware that is executing the code.
+
+.NET 7 introduced a set of new APIs for `Vector128<T>` and `Vector256<T>` for writing hardware-agnostic, cross platform vectorized code. The purpose of this document is to introduce you to the new APIs and provide a set of best practices.
+
+## Code structure
+
+`Vector128<T>` represents a 128-bit vector of type `T`. `T` is constrained to specific primitive types:
+
+* `byte` and `sbyte` (8 bits).
+* `short` and `ushort` (16 bits).
+* `int`, `uint` and `float` (32 bits).
+* `long`, `ulong` and `double` (64 bits).
+* `nint` and `unit` (32 or 64 bits, depending on the architecture)
+
+A single `Vector128` operation allows you to operate on: 16 (s)bytes, 8 (u)shorts, 4 (u)ints/floats, or 2 (u)longs/double(s).
+
+```
+------------------------------128-bits---------------------------
+|             64                |               64              |
+-----------------------------------------------------------------
+|      32       |      32       |      32       |      32       |
+----------------------------------------------------------------|
+|  16   |  16   |  16   |  16   |  16   |  16   |  16   |  16   |
+-----------------------------------------------------------------
+| 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 |
+-----------------------------------------------------------------
+```
+
+`Vector256<T>` is twice as big as `Vector128<T>`, so when it is hardware accelerated, and the data is large enough, you should use it instead of `Vector128<T>`. To check the acceleration, use `Vector128.IsHardwareAccelerated` and `Vector256.IsHardwareAccelerated` properties.
+
+The size of the input also matters. It needs to be at least of the size of a single vector to be able to execute the vectorized code path. `Vector128<T>.Count` and `Vector256<T>.Count` return the number of elements of the given type T in a single vector.
+Both APIs are turned into constants by the Just-In-Time compiler (i.e. no method call is required to retrieve the information). In the case of pre-compiled code (NativeAOT), this is not true for the `IsHardwareAccelerated` property, as the required information is not available at compile time.
+
+That is why the code is very often structured like this:
+
+```cs
+void CodeStructure(ReadOnlySpan<byte> buffer)
+{
+    if (Vector256.IsHardwareAccelerated && buffer.Length >= Vector256<byte>.Count)
+    {
+        // Vector256 code path
+    }
+    else if (Vector128.IsHardwareAccelerated && buffer.Length >= Vector128<byte>.Count)
```

Comment:
This will lead to two length comparisons for small data if both vectors are accelerated. For implementations that expect small inputs to be common, they'll often guard both sections with the 128 length check so that small data doesn't pay for both and you instead only pay for two checks if vectorization will happen, e.g.
```C#
void CodeStructure(ReadOnlySpan<byte> buffer)
{
    if (!Vector128.IsHardwareAccelerated || buffer.Length < Vector128<byte>.Count)
    {
        // scalar code path
    }
    else if (!Vector256.IsHardwareAccelerated || buffer.Length < Vector256<byte>.Count)
    {
        // Vector128 code path
    }
    else
    {
        // Vector256 code path
    }
}
```

Reply:
Out of curiosity: will PGO here re-arrange the conditions based on actual input-length present?